### PR TITLE
Example middleware that caches responses.

### PR DESCRIPTION
**Pull request by carld**

### CODE REVIEW

**Comment by carld** [comment]
@dakrone after enabling :debug true it looks like the network request is made regardless of the cache hit - can you see why?

**Comment by carld** [comment]
@dakrone this example "works for me", I have nothing more to add for now, if you have any suggestions on how to improve it, let me know. Thanks!

File: project.clj [hunk 1]
```diff
@@ -26,7 +26,8 @@
                                   [org.clojure/tools.logging "0.3.1"]
                                   [log4j "1.2.17"]
                                   [ring/ring-jetty-adapter "1.5.1"]
-                                  [ring/ring-devel "1.5.1"]]}
+                                  [ring/ring-devel "1.5.1"]
+                                  [org.clojure/core.cache "0.6.5"]]}
```
Comment (dakrone) [inline project.clj]: Can you add a comment next to this that it is only for the example?

File: examples/caching_middleware.clj [hunk 2]
```diff
@@ -0,0 +1,54 @@
+(ns clj-http.examples.caching-middleware
```
Comment (carld) [inline examples/caching_middleware.clj]: 👍

File: examples/caching_middleware.clj [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  (:require
+   [clj-http.client :as http]
+   [clojure.core.cache :as cache]))
+
+(def http-cache (atom (cache/ttl-cache-factory {} :ttl (* 60 60 1000))))
+
+(defn slurp-bytes
+  "Slurp the bytes from a slurpable thing"
+  [x]
+  (with-open [out (java.io.ByteArrayOutputStream.)]
```
Comment (carld) [inline examples/caching_middleware.clj]: Earlier when using slurp on it's own there were exceptions thrown when the response body was an gzipped http stream, which lead to a path of avoiding slurp altogether... I've used your suggestion and it is fine 👍

File: examples/caching_middleware.clj [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    (clojure.java.io/copy (clojure.java.io/input-stream x) out)
+    (.toByteArray out)))
+
+(defn- cached-response
+  ([client req]
+   (let [cache-key (str (:server-name req) (:uri req) "?" (:query-string req))]
+     (if (cache/has? @http-cache cache-key)
```
Comment (carld) [inline examples/caching_middleware.clj]: yes good catch

File: project.clj (hunk 1 above)
Comment (carld) [inline project.clj]: 👍

**Comment by dakrone** [comment]
Merged, thanks @carld!

**Pull request merged**